File: 4sumII.cpp

//Ian Hays
//08-26-2020
https://leetcode.com/problems/4sum-ii/
//SC: O(N^2) TC: O(N^2)

class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        int res = 0;
        unordered_map<int,int> umap;
        
        for(int i = 0; i < A.size(); i++){
            for(int j = 0; j < B.size(); j++){
                umap[A[i] + B[j]]++;
            }
        }
        for(int i = 0; i < C.size(); i++){
            for(int j = 0; j < D.size(); j++){
                int sum = -1 * (C[i] + D[j]); 
                if(umap[sum] != 0) res += umap[sum];  
            }
        }
        
        return res;
    }
};

File: Java/4sumII.java
//Ian Hays
//08-26-2020
//https://leetcode.com/problems/4sum-ii/
//SC: O(N^2) TC: O(N^2)

class Solution {
    public int fourSumCount(int[] A, int[] B, int[] C, int[] D) {
        int result = 0;
        HashMap<Integer,Integer> hmap = new HashMap<Integer,Integer>();
        
        for(int i = 0; i < A.length; i++){
            for(int j = 0; j < B.length; j++){
                int sum = A[i] + B[j];
                hmap.put(sum, hmap.getOrDefault(sum,0) + 1);
            }
        }
        
        for(int i = 0; i < C.length; i++){
            for(int j = 0; j < D.length; j++){
                int sum = C[i] + D[j];
                if(hmap.containsKey(-sum)) result += hmap.get(-sum);
            }
        }
        
        return result;
        
    }
}
